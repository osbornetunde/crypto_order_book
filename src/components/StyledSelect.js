import styled from "styled-components";

const StyledSelect = styled.select`
  display: block;
  font-weight: 400;
  line-height: 1.4;
  appearance: none;
  outline: 0;
  background: transparent;
  background-image: url("data:image/svg+xml;base64,PHN2ZyBhcmlhLWhpZGRlbj0idHJ1ZSIgZm9jdXNhYmxlPSJmYWxzZSIgZGF0YS1wcmVmaXg9ImZhcyIgZGF0YS1pY29uPSJjYXJldC1kb3duIiBjbGFzcz0ic3ZnLWlubGluZS0tZmEgZmEtY2FyZXQtZG93biBmYS13LTEwIiByb2xlPSJpbWciIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmlld0JveD0iMCAwIDMyMCA1MTIiPjxwYXRoIGZpbGw9ImN1cnJlbnRDb2xvciIgZD0iTTMxLjMgMTkyaDI1Ny4zYzE3LjggMCAyNi43IDIxLjUgMTQuMSAzNC4xTDE3NC4xIDM1NC44Yy03LjggNy44LTIwLjUgNy44LTI4LjMgMEwxNy4yIDIyNi4xQzQuNiAyMTMuNSAxMy41IDE5MiAzMS4zIDE5MnoiPjwvcGF0aD48L3N2Zz4="),
    linear-gradient(to bottom, #fff 0%, #e5e5e5 100%);
  background-repeat: no-repeat;
  background-position: right 0.7rem top 50%, 0 0;
  background-size: 0.65rem auto, 100%;
  border: 1px solid rgba(116, 223, 112, 0.9);
  width: 50%;
  height: 2rem;
  margin: 0;
  padding: 0 0 0 0.5rem;
  color: #000;
  cursor: pointer;
  border-radius: 0.5rem;
  box-shadow: 0 1px 0 1px rgba(0, 0, 0, 0.04);
  text-overflow: ellipsis;
  &:hover {
    border: 1px solid #888;
  }

  &:focus {
    border: 1px solid #000639;
    box-shadow: 0 0 1px 1px rgba(0, 6, 57, 0.3);
  }

  &::-ms-expand {
    display: none;
  }
`;

export default StyledSelect;
