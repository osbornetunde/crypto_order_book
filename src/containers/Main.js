import React, { useEffect } from "react";
import { connect } from "react-redux";
import { selectPair } from "../actions/pairsAction";
import {
  subscribe,
  fetchData,
  unSubscribe,
  stopStream,
  fetchStart,
  fetchEnd
} from "../actions/apiAction";
import StyledSelect from "../components/StyledSelect";
import StyledMainContainer from "../components/StyledMainContainer";
import StyledButton from "../components/StyledButton";

const Main = props => {
  const {
    pair,
    subscribe,
    unSubscribe,
    subscribeValue,
    unSubscribeValue,
    selectPair,
    fetchData,
    stopStream,
    fetchStart,
    fetchEnd
  } = props;

  useEffect(() => {
    const subscription = setTimeout(() => {
      subscribe(pair);
    }, 1000);
    return () => clearTimeout(subscription);
  }, [pair, subscribe]);

  useEffect(() => {
    const stoppingStream = setTimeout(() => {
      stopStream(unSubscribeValue);
    }, 1000);
    return () => clearTimeout(stoppingStream);
  }, [pair, stopStream, unSubscribeValue]);

  const setCurrentPair = e => {
    e.preventDefault();
    unSubscribe(pair);
    selectPair(e.target.value);
    fetchEnd(false);
  };

  const fetchDataHandler = () => {
    if (pair === "") {
      return null;
    }
    fetchStart(true);
    fetchData(subscribeValue);
  };

  const stopStreamHandler = () => {
    unSubscribe(pair);
    selectPair("");
    fetchEnd(false);
    stopStream(unSubscribeValue);
  };

  return (
    <StyledMainContainer>
      <StyledSelect value={pair} onChange={setCurrentPair}>
        <option value="" hidden>
          Select Currency Pair
        </option>
        <option value="bchbtc">BCH/BTC</option>
        <option value="bcheur">BCH/EUR</option>
        <option value="bchusd">BCH/USD</option>
        <option value="btceur">BTC/EUR</option>
        <option value="btcusd">BTC/USD</option>
        <option value="ethbtc">ETH/BTC</option>
        <option value="etheur">ETH/EUR</option>
        <option value="ethusd">ETH/USD</option>
        <option value="eurusd">EUR/USD</option>
        <option value="ltcbtc">LTC/BTC</option>
        <option value="ltceur">LTC/EUR</option>
        <option value="ltcusd">LTC/USD</option>
        <option value="xrpeur">XRP/EUR</option>
        <option value="xrpbtc">XRP/BTC</option>
        <option value="xrpusd">XRP/USD</option>
      </StyledSelect>
      <StyledButton onClick={fetchDataHandler}>Submit</StyledButton>
      <StyledButton onClick={stopStreamHandler} isStop isStopHover isStopFocus>
        Stop
      </StyledButton>
    </StyledMainContainer>
  );
};

const mapStateToProps = state => ({
  pair: state.pair,
  subscribeValue: state.subscribeValue,
  unSubscribeValue: state.unSubscribeValue
});

export default connect(
  mapStateToProps,
  {
    selectPair,
    subscribe,
    unSubscribe,
    fetchData,
    stopStream,
    fetchEnd,
    fetchStart
  }
)(Main);
