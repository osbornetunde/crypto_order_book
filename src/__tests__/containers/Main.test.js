import React from "react";
import { shallow } from "enzyme";
import { Provider } from "react-redux";
import { mockStore } from "../../mockStore/mockStore";
import Main from "../../containers/Main";
import StyledSelect from "../../components/StyledSelect";

StyledSelect.displayName = "StyledSelect";

let wrapper;
beforeEach(() => {
  wrapper = shallow(
    <Provider store={mockStore({})}>
      <Main />
    </Provider>
  ).find(Main);
});

describe("Main component", () => {
  it("Take a snapshot of the main component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("test for Main", () => {
    expect(wrapper.length).toEqual(1);
  });
  it("should show that the Main component exists", () => {
    expect(wrapper.exists()).toBe(true);
  });
});
