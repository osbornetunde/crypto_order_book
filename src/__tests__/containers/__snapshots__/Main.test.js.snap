// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component Take a snapshot of the main component 1`] = `ShallowWrapper {}`;
